doc/watermarks_HLD.md

```diff
@@ -0,0 +1,212 @@
+
+## 1.2 Modules description
+### 1.2.1 gRPC
+System data telemetry infrastructure. Basically allows to getRequest data from SONiC DBs (and more).
+### 1.2.2 Counter DB
+Located in the Redis DB instance #2 running inside the container "database". Redis DB works with the data in format of key-value tuples, needs no predefined schema and holds various counters like port counters, ACL counters, etc.
+### 1.2.3 Orchestration Agent
+This component is running in the "orchagent" docker container and is responsible for processing updates of the APP DB and do corresponding changes in the SAI DB via SAI Redis.
+### 1.2.4 SAI Redis
+SAI Redis is an implementation of the SAI API which translates API calls into SAI objects which are stored in the ASIC DB.
+### 1.2.5 ASIC DB
+Redis DB instance #1. Holds serialized SAI objects.
+### 1.2.6 syncd
+Reads SAI DB data (SAI objects) and performs appropriate calls to Switch SAI.
+### 1.2.7 SAI (Redis and Switch)
+An unified API which represent the switch state as a set of objects. In SONiC represented in two implementations - SAI DB frontend and ASIC SDK wrapper.
+# 2 Requirements
+The following watermarks should be supported:
+## 2.1 Watermark counters requirements
+|                                          | SAI attribute mapping  |
+|------------------------------------------|------------------------|
+| Ingress headroom per PG                  | SAI_INGRESS_PRIORITY_GROUP_STAT_XOFF_ROOM_WATERMARK_BYTES            |
+| Ingress shared pool occupancy per PG                | SAI_INGRESS_PRIORITY_GROUP_STAT_SHARED_WATERMARK_BYTES            |
+| Egress shared pool occupancy per queue (including both unicast queues and multicast queues)      | SAI_QUEUE_STAT_SHARED_WATERMARK_BYTES         |
+
+System behavior:
+We consider a maximum of one regular user and a maximum of one special user that comes from streaming telemetry (grpc)
+
+Streaming telemetry is only interested in periodic watermark, i.e., it queries the watermark at regular intervals. The interval is configurable. Streaming telemetry does not care about persistent watermark. 
+
+Regular user is able to query the watermark over a period specified in the cli. When the cli is issued, the cli blocks over the specified period, which counts the issuing of the cli as its starting point, and then returns the watermark over the specified period.
+Regular user is able to query the persistent watermark. Regular user is able to reset the persistent watermark. When the persistent watermark is reset, persistent watermark starts a new recording from the time reset is issued.
+
+When one regular user and the streaming telemetry coexist, they do not interfere with each other. Their behaviors stay the same as described above. So the software should be able to handle the following situations and return the correct watermark values to each user:
+
+![](https://github.com/mykolaf/SONiC/blob/gh-pages/images/watermark_HLD/timeline2.png)
```

Comment:
we need to walk through the timeline and connect these events to the detail design to see how everything works. currently, the description of events is not associated with your design.